fuzzer/src/main.rs

```diff
@@ -374,27 +418,35 @@ fn test_pattern(pattern: &str, time_samples: &mut [(f64, f64)]) -> PatternResult
     }
 }
 
-/// Returns the length of the tested substring and the time needed for parsing given string in
-/// given sample of given sample_size.
-///
-/// The passed string is the whole repeated pattern string.
-/// This function perform substring slicing according to the current sample and sample size uniformly.
-fn time_needed(repeated_pattern: &str, pattern_len: usize, sample: usize, sample_size: usize) -> (usize, Duration) {
-    let n = repeated_pattern.len() / sample_size * sample;
-    // round up to next pattern
-    let n = (n + pattern_len - 1) / pattern_len * pattern_len;
-
-    if DEBUG_LEVEL >= 3 {
-        println!("len: {}", n);
+/// Returns the number of repeated patterns.
+fn sample_pattern(
+    pattern: &Pattern,
+    buf: &mut String,
+    sample_size: usize,
+    sample_count: usize,
+) -> usize {
+    buf.clear();
+    buf.push_str(&pattern.prefix);
+
+    let target_byte_count = sample_size * NUM_BYTES / sample_count;
+    let target_repeat_bytes = target_byte_count - pattern.prefix.len() - pattern.postfix.len();
+    let num_repeats = target_repeat_bytes / pattern.repeating_pattern.len();
+
+    for _ in 0..num_repeats {
+        buf.push_str(&pattern.repeating_pattern);
     }
```

Comment:
The algorithm used by `str::repeat` (and `slice::repeat`) has an important optimization. It always memcopies the whole already repeated buffer as often as possible to provide exponential expansion and thus logarithmic time instead of linear time like here with the for-loop (<https://doc.rust-lang.org/src/alloc/slice.rs.html#434-487>).

Reply:
That's very interesting! Thanks for pointing it out. I'll try to rewrite this routine using those methods.

Reply:
It looks like `str::repeat` and `slice::repeat` don't quite work here since they themselves create owned data (`String`s) which is slow (O(n)) to prepend with the prefix.. These functions are also restricted to nightly for now.

I was able to optimize the pattern sampling function to continue building where the previous round left off. So it doesn't completely clear the buffer every time. This should make the string construction quite a bit faster.

Reply:
This sounds like a nice optimization, which should have a pretty huge impact. :+1: